tests/integration/src/test/java/org/mule/test/integration/client/MuleClientDispatchExceptionHandlingTestCase.java

```diff
@@ -0,0 +1,185 @@
+{
+
+    private static final int TIMEOUT = 5000;
+    private static final Log logger = LogFactory.getLog(MuleClientDispatchExceptionHandlingTestCase.class);
+
+    @Rule
+    public DynamicPort dynamicPort1 = new DynamicPort("port1");
+
+    private static CountDownLatch innerFlowLatch;
+    private static CountDownLatch exceptionLatch;
+
+    @Override
+    protected String getConfigFile()
+    {
+        return "org/mule/test/integration/client/client-dispatch-catch-exception-flow.xml";
+    }
+
+    /**
+     * Test to validate a dispatch event to a vm:inbound using MuleClient from a flow that has an http:listener.
+     * The catch-exception-strategy is called due to the JavaComponent throws an exception after doing the dispatch.
+     * Without the change to OutboundEnpointPropertyMessageProcessor the event was overwritten so default strategy
+     * listener was called instead.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testCatchExceptionThrowFromJavaComponentUsingHTTP() throws Exception
+    {
+        doSendMessageToEndpoint("http://localhost:" + dynamicPort1.getValue() + "/test", "Business Error!");
+    }
+
+    @Test
+    public void testCatchExceptionThrowFromMessageProcessorUsingVM() throws Exception
+    {
+        doSendMessageToEndpoint("vm://vmcatchexceptionfromprocessor", "Business Error!");
+    }
+
+    /**
+     * This tests validates that the event in threadLocal was overwritten when doing a dispatch from a JavaComponent
+     * and using MuleClient (or MuleEventContext) in order to do the dispatch to the sub-flow.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testCatchExceptionThrowFromJavaComponentUsingVM() throws Exception
+    {
+        doSendMessageToEndpoint("vm://vmcatchexception", "Business Error!");
+    }
+
+
+    /**
+     * On this scenario the call to the sub-flow was made with a vm connector (instead of using MuleClient) and it
+     * did work too without having the change OutboundPropertyMessageProcessor.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testCatchExceptionUsingVMConnector() throws Exception
+    {
+        doSendMessageToEndpoint("vm://vmconnector", "Business Error!");
+    }
+
+    private void doSendMessageToEndpoint(String endpoint, String expectedPayload) throws Exception
+    {
+        innerFlowLatch = new CountDownLatch(1);
+        exceptionLatch = new CountDownLatch(1);
+
+        LocalMuleClient client = muleContext.getClient();
+        final HttpRequestOptions httpRequestOptions = newOptions().method(POST.name()).responseTimeout(TIMEOUT).build();
+        logger.debug("Sending initial message to trigger execution of flow to be tested");
+        MuleMessage result = client.send(endpoint, getTestMuleMessage("Original Message"), httpRequestOptions);
+        logger.debug("Flow execution completed, result:" + result);
+
+        boolean innerFlowCalled = innerFlowLatch.await(3, TimeUnit.SECONDS);
```

Comment:
Use org.mule.tck.junit4.AbstractMuleContextTestCase#RECEIVE_TIMEOUT instead of magic number?